### ISSUE TITLE
Bundle that calls System.register multiple times

### ISSUE THREAD

**Issue by joeldenning** [rollup/rollup#2036]
I'm experimenting with replacing a bundle I currently create with `jspm bundle` for one created by rollup with system format. The purpose of my bundle is to register multiple named modules with SystemJS, all from one script tagged bundle. This preloads the modules so that they can be used by other systemjs modules without incurring additional network requests.

[Rollup repl example](https://rollupjs.org/repl?version=0.56.3&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMiUyRiolMjBUUkVFLVNIQUtJTkclMjAqJTJGJTVDbmltcG9ydCUyMCU3QiUyMGN1YmUlMjAlN0QlMjBmcm9tJTIwJy4lMkZtYXRocy5qcyclM0IlNUNuJTVDbmNvbnNvbGUubG9nKCUyMGN1YmUoJTIwNSUyMCklMjApJTNCJTIwJTJGJTJGJTIwMTI1JTIyJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMm1hdGhzLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMiUyRiUyRiUyMG1hdGhzLmpzJTVDbiU1Q24lMkYlMkYlMjBUaGlzJTIwZnVuY3Rpb24lMjBpc24ndCUyMHVzZWQlMjBhbnl3aGVyZSUyQyUyMHNvJTVDbiUyRiUyRiUyMFJvbGx1cCUyMGV4Y2x1ZGVzJTIwaXQlMjBmcm9tJTIwdGhlJTIwYnVuZGxlLi4uJTVDbmV4cG9ydCUyMGZ1bmN0aW9uJTIwc3F1YXJlJTIwKCUyMHglMjApJTIwJTdCJTVDbiU1Q3RyZXR1cm4lMjB4JTIwKiUyMHglM0IlNUNuJTdEJTVDbiU1Q24lMkYlMkYlMjBUaGlzJTIwZnVuY3Rpb24lMjBnZXRzJTIwaW5jbHVkZWQlNUNuZXhwb3J0JTIwZnVuY3Rpb24lMjBjdWJlJTIwKCUyMHglMjApJTIwJTdCJTVDbiU1Q3QlMkYlMkYlMjByZXdyaXRlJTIwdGhpcyUyMGFzJTIwJTYwc3F1YXJlKCUyMHglMjApJTIwKiUyMHglNjAlNUNuJTVDdCUyRiUyRiUyMGFuZCUyMHNlZSUyMHdoYXQlMjBoYXBwZW5zISU1Q24lNUN0cmV0dXJuJTIweCUyMColMjB4JTIwKiUyMHglM0IlNUNuJTdEJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMnN5c3RlbSUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE) (Notice that there is only one System.register)

Currently rollup does not allow bundles outputted to system format to be named at all, but see #2028 for relevant pr that adds this.

Even once that pr is merged, though, there still will not be (afaik) a way to create a bundle that calls System.register multiple times. What are your thoughts on how such a bundle would best be created? Here are the things I have tried and my current options:

## Option 1 - multiple rollup configs with a separate concat process to combine the outputted bundles
```js
// Each object represents something I want to 
module.exports = [{...}, {...}]
```
This is actually what I'm doing right now as a workaround -- I use `cat bundle1.js bundle2.js ... > master-bundle.js` after rollup creates ~15 separate bundles. The major drawback is that `cat` doesn't preserve source maps.

## Option 2 - use rollup js api (currently not supported afaik)
```js
Promise.all(rollup(inputConfig1), rollup(inputConfig2)).then(bundles => {
  const [bundle1, bundle2] = bundles;
  // ???? Combine the bundles somehow before calling masterBundle.writeFile() ????
})
```

## Option 3 - use multi-entry plugin (doesn't quite do what I want afaict)
[rollup-plugin-multi-entry](https://github.com/rollup/rollup-plugin-multi-entry) allows you to specify multiple entries, which sounds promising. However it still produces just one bundle and, in turn, one System.register call. So this one doesn't work

## Option 4 - somehow configure rollup to System.register multiple things in the output bundle
I'm not sure exactly what this would look like, but perhaps with some work in the rollup codebase something like this could produce the desired bundle:
```js
import 'react' // First System.register call
import 'lodash' // Second System.register call
```

## Option 5 - maybe rollup shouldn't support this at all
Let me know if what I'm trying to do with rollup is way outside the bounds of what it is meant for! I am experimenting with rollup right now before the company I work for fully embraces it, but we can stick to `jspm bundle` if this use case is not very easily supported.

**Comment by joeldenning** [comment]
@guybedford thoughts?

**Comment by guybedford** [comment]
@joeldenning I'd probably suggest to firstly use the code splitting output of Rollup here, and then secondly to run something like gulp-concat on the output dist folder to turn everything into a single bundle. Alternatively custom bundles can be constructed from dependency tracing.

It's not necessarily obvious what the best bundle groupings would be for code splitting as well. It would help to hear what sort of approaches you are using to share bundles here in a little more detail if that sounds like it could be useful to discuss?

**Comment by joeldenning** [comment]
> It would help to hear what sort of approaches you are using to share bundles here in a little more detail

We create a bundle "common-dependencies.prod.js" that registers react, react-dom, prop-types, lodash, moment, rx, etc to SystemJS. Then we do `SystemJS.import('app1')`, which is in an AMD format and has dependencies on the common dependencies. The common dependencies and app1 are deployed separately, which is why it's nice for SystemJS to be our module registry and do the dependency injection at runtime in the browser. Currently we're using `jspm bundle react + react-dom ...` to do this, but are thinking that rollup might let us fine tune the size and performance of the bundle. The motivations for trying out rollup's system format was (1) that we more easily control exactly which modules are registered and which ones aren't (mostly as a potential performance improvement -- `jspm bundle` registers more than just the entry files, which we don't actually need), (2) we can use yarn and node_modules which is is more convenient and familiar for our developers, and (3) we could fine tune or even eliminate our config.js file that calls SystemJS.config() with all the npm/github packages and resolution, and maybe explore using system.production.js instead of system.js.

As I've dived into using rollup for our common dependencies bundle, I've become reacquainted with some of the nitty gritty things that jspm hides from us, like `__esModule` for interop with webpack and deciding what to do with commonjs' module.exports. The more I've gotten into it, the more I'm thinking that maybe we could get the bundle to be smaller and more performant simply by being careful to bundle just the `.min.js` files that are published by most of the npm packages. And that could be done with `jspm bundle react/umd/react.min.js`, which would not require rollup.

So we're really sort of just in experimentation mode -- trying to figure out which approach would be best for us in the long term.

**Comment by joeldenning** [comment]
Oh another issue with how we're doing our bundle with `jspm bundle` right now is that afaik we can't tell it to replace `process.env.NODE_ENV` and then do dead code elimination. So `jspm bundle react` results in both the development and prod versions of react being bundled. That might just be user error on our part, but we're thinking with rollup we'll be able to use the replace plugin and uglify to do dead code elimination on the version of react that we don't care about.

**Comment by guybedford** [comment]
> but are thinking that rollup might let us fine tune the size and performance of the bundle. The motivations for trying out rollup's system format was (1) that we more easily control exactly which modules are registered and which ones aren't (mostly as a potential performance improvement -- jspm bundle registers more than just the entry files, which we don't actually need), (2) we can use yarn and node_modules which is is more convenient and familiar for our developers, and (3) we could fine tune or even eliminate our config.js file that calls SystemJS.config() with all the npm/github packages and resolution, and maybe explore using system.production.js instead of system.js.

These are all great reasons. I'd strongly suggest trying out the code splitting workflow in Rollup here as it is designed exactly to create the minimum number of registered modules that can then be grouped into named bundle files as necessary.

Rollup plugins should be able to provide the process.env.NODE_ENV and CommonJS interpretation, it's just about getting the right configuration in place.

**Comment by joeldenning** [comment]
[Your comment](https://github.com/rollup/rollup/pull/2028#issuecomment-372993312) that talks about potentially adding a `combineNamedModules` option would be awesome for us -- it would resolve this issue in a very clean way. You mentioned there that you have a wip branch to do just that -- will it include the combineNamedModules option? Is there more discussion to be had before code splitting and named modules will interact?

**Comment by guybedford** [comment]
@joeldenning I've responded at https://github.com/rollup/rollup/pull/2028#pullrequestreview-105326453.

**Comment by guybedford** [comment]
I don't think this is would be suitable for the Rollup approach to bundles as single shared scopes, as discussed.
